### ISSUE TITLE
[Containerd] Disabling IPv6 

### ISSUE THREAD

**Issue by xuoja-acc** [aws/amazon-vpc-cni-k8s#2483]
**What happened**: 
After upgrade to kubernetes 1.24 and containerd pods resolve localhost on the ipv6 address ::1.
The upgrade documentation warn us about this behavior but the documentation doesn't provides the way to disable IPv6 at **the instance level**. 

We tried to disable the IPv6 through the bootstrap + sysctl but containers are still created with IPv6 (reboot is **not** an option but we tried with no effects)
```
cat <<'EOF' > /etc/sysctl.d/10-disable-ipv6.conf
net.ipv6.conf.all.disable_ipv6 = 1
net.ipv6.conf.default.disable_ipv6 = 1
net.ipv6.conf.lo.disable_ipv6 = 1
EOF
```

**What you expected**:
We shoul'd be able to disable the ipv6 by configuring CNI with the[ **cni tuning** plugin ](https://www.cni.dev/plugins/current/meta/tuning/) or any other way.

**NB** : Adding the tuning part to nerdctl-bridge.conflist is perfectly working

```
{
  "cniVersion": "0.4.0",
  "name": "aws-cni",
  "disableCheck": true,
  "plugins": [
    {
      "name": "aws-cni",
      "type": "aws-cni",
      "vethPrefix": "eni",
      "mtu": "9001",
      "podSGEnforcingMode": "strict",
      "pluginLogFile": "/var/log/aws-routed-eni/plugin.log",
      "pluginLogLevel": "DEBUG"
    },
    {
      "name": "egress-cni",
      "type": "egress-cni",
      "mtu": "9001",
      "enabled": "false",
      "randomizeSNAT": "prng",
      "nodeIP": "",
      "ipam": {
         "type": "host-local",
         "ranges": [[{"subnet": "fd00::ac:00/118"}]],
         "routes": [{"dst": "::/0"}],
         "dataDir": "/run/cni/v4pd/egress-v6-ipam"
      },
      "pluginLogFile": "/var/log/aws-routed-eni/egress-v6-plugin.log",
      "pluginLogLevel": "DEBUG"
    },
    {
      "type": "portmap",
      "capabilities": {"portMappings": true},
      "snat": true
    },
    {
      "type": "tuning",
      "sysctl": {
          "net.ipv6.conf.all.disable_ipv6": "1",
          "net.ipv6.conf.default.disable_ipv6": "1",
          "net.ipv6.conf.lo.disable_ipv6": "1"
       }
    }
  ]
}
```

NB : adding the tuning part to nerdctl-bridge.conflist is perfectly working

**[AWS Documentation](https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html)**: 
```
Containerd has IPv6 enabled for Pods, by default. It applies node kernel settings to Pod network namespaces. Because of this, containers in a Pod bind to both IPv4 (127.0.0.1) and IPv6 (::1) loopback addresses. IPv6 is the default protocol for communication. Before updating your cluster to version 1.24, we recommend that you test your multi-container Pods. Modify apps so that they can bind to all IP addresses on loopback interfaces. The majority of libraries enable IPv6 binding, which is backward compatible with IPv4. When it’s not possible to modify your application code, you have two options:

Run an init container and set disable ipv6 to true (sysctl -w net.ipv6.conf.all.disable ipv6=1).

Configure a mutating admission webhook to inject an init container alongside your application Pods.

If you need to block IPv6 for all Pods across all nodes, you might have to disable IPv6 on your instances.
```

**Environment**:
- Kubernetes version : `1.25`
- CNI Version : `v1.13.2`
- OS : `Amazon Linux 2`
- Kernel : `Linux runner-e12a2eg-project-418-concurrent-0phzqn 5.10.184-175.749.amzn2.x86_64 #1 SMP Wed Jul 12 18:40:28 UTC 2023 x86_64 Linux`

**Comment by jdn5126** [comment]
@xuoja-acc can you link to the documentation you are referring to? This issue is not specific to the VPC CNI, as this is a node-level configuration. I would expect this to be taken care of by the AMI, and I am not aware of containerd globally enabling IPv6

**Comment by xuoja-acc** [comment]
Hi @jdn5126 , here is the link of the documentation mentioning informations about ipv6 : https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html

**Disabling this options at the OS level doesn't have effects and containers are still created with ipv6 **  : 
```
          "net.ipv6.conf.all.disable_ipv6": "1",
          "net.ipv6.conf.default.disable_ipv6": "1",
          "net.ipv6.conf.lo.disable_ipv6": "1"
```

**Comment by jdn5126** [comment]
Currently, the VPC CNI only enables IPv6 if `ENABLE_IPv6` is set and IPv6 is not already enabled in the kernel: https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/aws-vpc-cni-init/main.go#L103 .

Changing that default to disable IPv6 if `ENABLE_IPv6` is false is not an option, and I feel adding another environment variable here is a bad idea. If you set `net/ipv6/conf/all/disable_ipv6` to 1, then IPv6 will be disabled system-wide. That won't prevent a loopback interface from getting created in the container with `::1`, though. 

I am confused by the comment about the `tuning` plugin as everything configured at the system-level will be inherited by new network namespaces. Regardless, customers are still free to use the `tuning` plugin, we don't prevent that.

**Comment by xuoja-acc** [comment]
Currently : 
- Our userdata script set `net/ipv6/conf/all/disable_ipv6` to 1, IPv6 is be disabled system-wide and we already have  `ENABLE_IPv6`  set to  false.
- In the userdata, the script set `disable_ipv6` before we run the /etc/eks/bootstrap.sh : so the IPv6 is disabled system-wide before any kubernetes components run.

That won't prevent a loopback interface from getting ipv6 loopback, **and also getting ipv6 on the pod interface** :

```
/ # ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
3: eth0@if10: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 9001 qdisc noqueue state UP 
    link/ether 62:5e:ed:38:37:29 brd ff:ff:ff:ff:ff:ff
    inet 100.64.224.164/32 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::605e:edff:fe38:3729/64 scope link
       valid_lft forever preferred_lft forever
```

I manually added the tuning conf to `/etc/cni/net.d/10-aws.conflist`, and this results as disabling completly IPv6 :
```
/ # ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
3: eth0@if11: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 9001 qdisc noqueue state UP 
    link/ether 22:cf:2d:9d:3e:3c brd ff:ff:ff:ff:ff:ff
    inet 100.64.224.73/32 scope global eth0
       valid_lft forever preferred_lft forever
```

`Regardless, customers are still free to use the tuning plugin, we don't prevent that.`
I'm sorry but i don't get how to add **automatically** the tuning part in the 10-aws.conflist . 
Is there a way to add it "cni" widly ? A way to merge something like /etc/cni/net.d/00-disable-ipv6-.conflist + 10-aws.conflist

**Comment by jdn5126** [comment]
Ah, I see your point now. Since the `aws-node` pod installs the conflist fresh each time, you do not have a programmatic way to insert the `tuning` plugin other than manually creating a higher priority conflist in `/etc/cni/net.d/`, i.e. `/etc/cni/net.d/0-aws_tuning.conflist`.

I labeled this as a feature request as we'll have to think of a good way to solve this. The loopback interface is created by `containerd`, so we do not have much flexibility here. As a side note, is the presence of the IPv6 address on the loopback address causing you any application issues?

**Comment by xuoja-acc** [comment]
You are right, adding tuning through a programmatic way is exactly what we need ! absolutely

You are absolutely right : adding tuning through a programmatic way is exactly what we need !

For example the IPv6 loopback address is causing issues using localstack   :  `AWS SDK error wrapper for Error: connect ECONNREFUSED ::1:4566`

https://docs.localstack.cloud/references/network-troubleshooting/ :
`LocalStack only binds to IPv4 addresses (e.g. 127.0.0.1). Check you are not trying to access LocalStack over IPv6.`

**Comment by jdn5126** [comment]
Got it, I have marked this as a feature request and will work with our Project Management team to figure out how to prioritize this. In the meantime, can you disable LocalStack from trying to bind to to the IPv6 address?

**Comment by jdn5126** [comment]
As a side note, I created a new conflist on each node called `/etc/cni/net.d/1-aws_tuning.conflist` and verified that it takes precedence over the default installed `10-aws.conflist` and persists on reload. So that can serve as a good workaround in the meantime.

**Comment by xuoja-acc** [comment]
Thank's for the side note !

Another word about this topic : I wasn't aware of an existing chart but it may be a solution to overide the 10-aws.conflist throught the [cniConfig.fileContents](https://github.com/aws/amazon-vpc-cni-k8s/blob/master/charts/aws-vpc-cni/values.yaml#L65)

**Comment by jdn5126** [comment]
> Thank's for the side note !
> 
> Another word about this topic : I wasn't aware of an existing chart but it may be a solution to overide the 10-aws.conflist throught the [cniConfig.fileContents](https://github.com/aws/amazon-vpc-cni-k8s/blob/master/charts/aws-vpc-cni/values.yaml#L65)

I think there may have been plans to use that at some point, but currently nothing consumes from the ConfigMap that writes to: https://github.com/aws/amazon-vpc-cni-k8s/blob/master/charts/aws-vpc-cni/templates/configmap.yaml#L8

The current plan is for us to implement this via an environment variable, i.e. something like `TUNING_SYSCTLS`, but we are still evaluating

**Comment by mattburgess** [comment]
@jdn5126 we're running into this as well and can't work out why the `disable_ipv6` sysctl values aren't being inherited by pods correctly. They do when running under the dockershim runtime.

Our particular failure case is that pod health checks that try to connect to `localhost` see that name resolve to an ipv6 address, but the service is only listening on the ipv4 address hence the healthcheck fails with a `connection refused` error.

Could you provide the contents of the `/etc/cni/net.d/1-aws_tuning.conflist` file you used to test the precedence please? I've succeeded in getting the following to work, but I'm unsure as to whether or not the configuration of the `egress-v4-cni` plugin in `10-aws.conflist` is going to be taken into account (i.e do the 2 configs get merged?). I'm wary of writing that out myself as it has node-specific/subnet-specific info in it so can't be baked into our AMIs:

```
{
  "cniVersion": "0.4.0",
  "name": "aws-cni",
  "disableCheck": true,
  "plugins": [
    {
      "name": "aws-cni",
      "type": "aws-cni",
      "vethPrefix": "eni",
      "mtu": "9001",
      "podSGEnforcingMode": "strict",
      "pluginLogFile": "/var/log/aws-routed-eni/plugin.log",
      "pluginLogLevel": "DEBUG"
    },
    {
      "type": "tuning",
      "sysctl": {
        "net.ipv6.conf.all.disable_ipv6": "1",
        "net.ipv6.conf.default.disable_ipv6": "1"
      }
    }
  ]
}
```

**Comment by jdn5126** [comment]
@mattburgess containerd ignores the host-level networking sysctls and creates container networking namespaces with IPv6 enabled by default. Why it does this, and why it does not make this behavior configurable, is not clear. They appear to be making the statement that it is the application's responsibility (and the health check's responsibility) to bind only to the address family that it needs to use.

For the conflist, you can ignore `egress-cni`, as it is only enabled for IPv6 clusters or when explicitly set. Note the `"enabled": "false"` field.

The contents of the `1-aws_tuning.conflist` file that I used with VPC CNI v1.13.0+ is:
```
{
  "cniVersion": "0.4.0",
  "name": "aws-cni",
  "disableCheck": true,
  "plugins": [
    {
      "name": "aws-cni",
      "type": "aws-cni",
      "vethPrefix": "eni",
      "mtu": "9001",
      "podSGEnforcingMode": "strict",
      "pluginLogFile": "/var/log/aws-routed-eni/plugin.log",
      "pluginLogLevel": "DEBUG"
    },
    {
      "name": "egress-cni",
      "type": "egress-cni",
      "mtu": "9001",
      "enabled": "false",
      "randomizeSNAT": "prng",
      "nodeIP": "",
      "ipam": {
         "type": "host-local",
         "ranges": [[{"subnet": "fd00::ac:00/118"}]],
         "routes": [{"dst": "::/0"}],
         "dataDir": "/run/cni/v4pd/egress-v6-ipam"
      },
      "pluginLogFile": "/var/log/aws-routed-eni/egress-v6-plugin.log",
      "pluginLogLevel": "DEBUG"
    },
    {
      "type": "portmap",
      "capabilities": {"portMappings": true},
      "snat": true
    },
    {
      "type": "tuning",
      "sysctl": {
          "net.ipv6.conf.all.disable_ipv6": "1",
          "net.ipv6.conf.default.disable_ipv6": "1",
          "net.ipv6.conf.lo.disable_ipv6": "1"
       }
    }
  ]
}
```

**Comment by mattburgess** [comment]
Doh! Should really have spotted that `"enabled": "false"` line earlier. That's perfect, thanks! Looking forward to that `TUNING_SYSCTLS` setting (or similar), but this gets us unblocked for now. Thanks for the quick feedback!

**Comment by mattburgess** [comment]
> They appear to be making the statement that it is the application's responsibility (and the health check's responsibility) to bind only to the address family that it needs to use.

Just a slight nitpick, but this is what we _are_ doing. The problem is that because the pods have IPv6 enabled, that affects DNS resolution; resolvers return AAAA records before A records and because kubelet populates the pod's `/etc/hosts` file with IPv6 and IPv4 entries for `localhost` (again, regardless of IPv6 being enabled on the host or not, and again not configurable), then DNS clients see `::1` returned. The health check then fails because the app isn't bound to `::1`.

I'd normally go and raise an issue over at the containerd folks but a) I doubt they'll change this behaviour now and b) it'll take ages for any fix that they might make to get into our distro of choice and is unlikely to be backported to our distro version anyway.

**Comment by jdn5126** [comment]
That's a great call out. I am surprised that we have not seen more customer issues related to this containerd behavior, but perhaps we will as the EKS 1.23 EOL approachs.

**Comment by mattburgess** [comment]
Yeah, I was tearing my hair out when I couldn't find any meaningful google results! I've been a good netizen though and reported https://github.com/containerd/containerd/issues/8958 so even if it doesn't get any upstream traction folks might at least find solace in the fact that they're not alone!

**Comment by jdn5126** [comment]
https://github.com/aws/amazon-vpc-cni-k8s/pull/2499 has merged and will ship in `v1.15.0`, but thanks to @henry118, we were able to determine that there is another path: disabling IPv6 in the kernel bootloader.

System services on Linux, like `containerd`, inherit the bootloader settings when creating new network namespaces (https://www.mail-archive.com/debian-ipv6@lists.debian.org/msg04454.html).

I confirmed that the following disables IPv6 at the host level and in new network namespaces:
1. Add `ipv6.disable=1` to `GRUB_CMDLINE_LINUX_DEFAULT` in `/etc/default/grub`
2. Generate new grub config with `grub2-mkconfig -o /boot/grub2/grub.cfg`
3. Reload node

Note that for EKS AMIs, you can do the following:
1. Launch an EC2 instance.
2. Make the above to `grub`
3. Stop the EC2 instance.
4. Call `EC2::CreateImage` to create an AMI from the point-in-time EBS snapshot: https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateImage.html

**Comment by VLZZZ** [comment]
> > Thank's for the side note !
> > Another word about this topic : I wasn't aware of an existing chart but it may be a solution to overide the 10-aws.conflist throught the [cniConfig.fileContents](https://github.com/aws/amazon-vpc-cni-k8s/blob/master/charts/aws-vpc-cni/values.yaml#L65)
> 
> I think there may have been plans to use that at some point, but currently nothing consumes from the ConfigMap that writes to: https://github.com/aws/amazon-vpc-cni-k8s/blob/master/charts/aws-vpc-cni/templates/configmap.yaml#L8
> 
> The current plan is for us to implement this via an environment variable, i.e. something like `TUNING_SYSCTLS`, but we are still evaluating

I think it should be consumed as indented because it's being [mounted to the /app/10-aws.conflist](https://github.com/aws/amazon-vpc-cni-k8s/blob/4503f75ca0b60dd0f67f3ad9dfbddacffaa97282/charts/aws-vpc-cni/templates/daemonset.yaml#L103C1-L109C11):
```
{{- if .Values.cniConfig.enabled }}
            # The dockerfile copies the baked in config to this location, so overwrite it with ours.
            # The entrypoint process will then copy our config to /host/etc/cni/net.d on boot.
          - name: cni-config
            mountPath: /app/10-aws.conflist
            subPath: 10-aws.conflist
{{- end }}
```

And then it's should be correctly handled [due of this const](https://github.com/aws/amazon-vpc-cni-k8s/blob/4503f75ca0b60dd0f67f3ad9dfbddacffaa97282/cmd/aws-vpc-cni/main.go#L66):
```
defaultAWSconflistFile       = "/app/10-aws.conflist"
```

Or am I missign something? (Except of this bug - #2551)  
Just trying to mitigate the IPv6 disabling with as low effor as possible.

**Comment by jdn5126** [comment]
Closing since this support shipped in v1.15.0
